File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.java
class Solution {
    public int findPairs(int[] nums, int k) {
        Arrays.sort(nums);
        int cnt = 0; HashSet<Integer>set = new HashSet<>();

        for(int i = 0; i < nums.length; i++){
            int x = nums[i];
            if(set.contains(x))
                continue;
            set.add(x);
            if(Arrays.binarySearch(nums, i+1, nums.length, x+k) >= 0){
                cnt += 1;
            }
        }   

        return cnt;
    }
}

File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.cpp
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        //this can also be solved using running pointer and prefix array
        int n = nums.size();
        map<int, int>map;
        int cnt = 0;
        set<pair<int,int>>set;
        for(int i = 0; i < n; i++){
            if(map[k+nums[i]] > 0){
                int f,s;
                if(k+nums[i] < nums[i]){
                    f = k+nums[i];
                    s = nums[i];
                } else {
                    f = nums[i];
                    s = k+nums[i];
                }
                set.insert({f,s});
            }
            if(map[-k+nums[i]] > 0){
                int f,s;
                if(-k+nums[i] < nums[i]){
                    f = -k+nums[i];
                    s = nums[i];
                } else {
                    f = nums[i];
                    s = -k+nums[i];
                }
                set.insert({f,s});
            }
            map[nums[i]]++;
        }

        return set.size();
    }
};